test-suite/exonum-py-tests/exonum_tests/migrate.py

```diff
@@ -0,0 +1,280 @@
+        )
+        resume_config = Configuration(resume_config_dict)
+
+        with Launcher(resume_config) as launcher:
+            launcher.start_all()
+            launcher.wait_for_start()
+
+            self.wait_for_api_restart()
+            # Check that the service status has been changed to `active`.
+            for service in client.public_api.available_services().json()["services"]:
+                if service["spec"]["name"] == INSTANCE_NAME:
+                    self.assertEqual(service["status"]["type"], "active")
+                    self.assertEqual(service["spec"]["artifact"]["version"], "0.2.0")
+
+        # Unload artifact with version 0.1.0
+        unload_config_dict = generate_config(
+            self.network, instances=instances, artifact_action="unload", artifact_version="0.1.0"
+        )
+        unload_config = Configuration(unload_config_dict)
+
+        with Launcher(unload_config) as launcher:
+            launcher.unload_all()
+            launcher.wait_for_unload()
+
+            self.wait_for_api_restart()
+            explorer = launcher.explorer()
+
+            for artifact in unload_config.artifacts.values():
+                deployed = explorer.is_deployed(artifact)
+                self.assertEqual(deployed, False)
+
+    def test_migrate_running_service(self):
+        """Tests migration flow when the migrating service is running."""
+
+        # Deploy a service with 0.2.0 version.
+        instances = {INSTANCE_NAME: {"artifact": "cryptocurrency"}}
+        config_dict = generate_config(self.network, instances=instances)
+        deploy_config = Configuration(config_dict)
+
+        with Launcher(deploy_config) as launcher:
+            launcher.deploy_all()
+            launcher.wait_for_deploy()
+
+            self.wait_for_api_restart()
+            explorer = launcher.explorer()
+            for artifact in launcher.launch_state.completed_deployments():
+                deployed = explorer.is_deployed(artifact)
+                self.assertEqual(deployed, True)
+
+        # Migrate service data from 0.1.0 to 0.2.0 version
+        migrations = {INSTANCE_NAME: {"runtime": "rust", "name": "exonum-cryptocurrency", "version": "0.2.0"}}
+        migrations_dict = generate_migration_config(self.network, migrations)
+        migration_config = Configuration(migrations_dict)
+
+        with Launcher(migration_config) as launcher:
+            launcher.migrate_all()
+            launcher.wait_for_migration()
+
+            for instance, (status, message) in launcher.launch_state.completed_migrations().items():
+                if instance == INSTANCE_NAME:
+                    self.assertEqual(status, ActionResult.Fail)
+                    self.assertTrue("is not stopped or frozen" in message)
+
+    def test_migrate_without_migration_logic_stage(self):
+        """Tests migration flow without migration logic stage."""
+
+        host, public_port, private_port = self.network.api_address(0)
```

Comment:
Is it possible to de-duplicate the code from here and `test_full_migration_flow`?